gpcontrib/arenadata_toolkit/arenadata_toolkit.c

```diff
@@ -1,4 +1,198 @@
+#include <sys/stat.h>
+
 #include "postgres.h"
+
+#include "access/aomd.h"
+#include "access/heapam.h"
+#include "cdb/cdbvars.h"
+#include "common/relpath.h"
 #include "fmgr.h"
+#include "miscadmin.h"
+#include "storage/lock.h"
+#include "utils/builtins.h"
+#include "utils/elog.h"
+#include "utils/rel.h"
+#include "utils/relcache.h"
 
 PG_MODULE_MAGIC;
+
+/*
+ * MAXPATHLEN_WITHSEGNO: size of path buffer for relation segment
+ * ('segno' may require 12 additional bytes).
+ */
+#define MAXPATHLEN_WITHSEGNO (MAXPGPATH + 12)
+
+static int64 calculate_relation_size(Relation rel, ForkNumber forknum);
+static int64 get_heap_storage_total_bytes(Relation rel,
+							 ForkNumber forknum, char *relpath);
+static int64 get_ao_storage_total_bytes(Relation rel, char *relpath);
+static bool calculate_ao_storage_perSegFile(const int segno, void *ctx);
+static void fill_relation_seg_path(char *buf, int bufLen,
+					   const char *relpath, int segNo);
+
+/*
+ * Structure used to accumulate the size of AO/CO relation from callback.
+ */
+struct calculate_ao_storage_callback_ctx
+{
+	char	   *relfilenode_path;
+	int64		total_size;
+};
+
+/*
+ * Function to calculate size of a relation by it's OID and optional forkNumber
+ * (by default it's MAIN). The implementation of function is based on
+ * pg_relation_size from dbsize.c
```

Comment:
What is the difference between them?

Reply:
There is only one difference - in text of sql string.